### ISSUE TITLE
Don't apply formatting in the raw Markdown editor when Markdown preview is enabled

### ISSUE THREAD

**Issue by gparyani** [StackExchange/Stacks-Editor#236]
I like how the formatting of raw Markdown text in the editor changes based on what's typed in, such as showing text formatted with italic markup as italicized, when I've chosen the option to only show the raw Markdown. If I'm only typing raw Markdown with no preview, it makes sense to provide a preliminary indication as to how something will look in the editor.

However, I find this redundant and annoying when I've selected the option to show the preview pane. I can already see how my text is coming out in the preview pane; I don't need to be told again how it will be like in the editor.

It's also really annoying as I'm typing to see my text change as I type it. As an example, when I'm typing a heading using the format of putting one or more dashes after a line of text, the font size change in the editor causes my cursor to change position, which is unexpected. It's also inconsistent, as text formatted using HTML won't show as formatted in the editor.

Finally, this introduces accessibility concerns. As an example, when editing code, or when editing some part of the post that doesn't support formatting, the text appears grayer, which reduces contrast ratio and makes the text harder to read. Another example is for italic text, in case the user's system monospace font is one that's hard to read when italicized (some such fonts exist).

Another annoyance that may come up is if the editor fails to recognize some section of the post as part of a block that doesn't support formatting: the text will show as formatted in the editor, but not in the preview or actual post. The team has done a good job of sniffing those out and I haven't run into this in the cases I've thought of (spoilers and multi-line `<sub>` or `<sup>` blocks), but it could come up in the future.

To summarize, please don't format the Markdown in the editor when the preview is enabled. It's useful when there's no preview, but when there is, it's redundant and annoying. Better yet, provide an option to enable/disable formatting in the Markdown editor, independent of whether the preview pane is enabled or not.

**Comment by b-kelly** [comment]
It looks like you're requesting changes to two distinct classes of behavior here. I'll address them individually below:

# Remove commonmark syntax highlighting when the preview is shown

We currently do not have any plans for disabling the syntax highlighting in commonmark mode. Syntax highlighting isn't meant to provide a preview of what the rendered text will look like, but rather to provide color prompts to the writer (and potentially, reader down the line) to help break up the document into visually distinct parts. In this case, even though there's a rendered preview available, it doesn't actually assist in the authoring experience in the same way that syntax highlighting does. I'll keep my eyes open for other requests like this, but for now we haven't heard from others that the syntax highlighting is annoying when the preview is present.

# UX and accessibility concerns for syntax highlighting

Accessibility first, as that's easier to objectively address: Regarding the gray font color used for code (and html), it passes WCAG AAA compliance for _small_ text (7.10:1 in light mode, 8.57:1 in dark mode), which well exceeds the minimum required for accessible contrast in text. In regards to poor italic system fonts, I'm not sure what all we can do there. I'd recommend reaching out to the entity in charge of setting that font on your OS and report your concerns to them.

Regarding text changing as you type it: Is this primarily a concern with the line height / font size of headings, or does it also include color changing as syntax is being recognized? For headings, we may consider simply consolidating the size there to prevent the cursor from jumping. If that's something you feel strongly about, I'd recommend opening an issue focusing solely on that.

Regarding blocks that are incorrectly highlighted: Do you have any specific examples of this? If a block is not being properly syntax highlighted, that would be considered a bug. There _are_ a few syntaxes that we're aware of not being highlighted at all, but these are solely constrained to Stack Exchange specific syntax additions (e.g. tags, spoilers).

---

Thank you for your feedback and report! I'm closing this issue for now, as there are currently no actionable items. I'm happy to reopen if you have any questions, or if new actionable items are discovered through discourse.

**Comment by gparyani** [comment]
@b-kelly Thanks for your response!

With regard to the first heading, I personally find the CommonMark syntax highlighting extremely annoying and would personally prefer that it be off in all cases, whether the preview is showing or not. I understand the positives of highlighting the text, but it would be nice to provide an option to disable the highlighting if one finds it undesired. I note this in the final paragraph of my original report, in the sentence beginning with "Better yet".

I will admit that my judgment was a bit clouded here - my eyes were primarily focused on the relative comparison between heading text and normal text in the Stacks editor, and I didn't notice that much of my issue was more to do with the text being smaller between new and old editors on the Stack Exchange sites than with the fact that the text was being highlighted - I've filed [a request on Meta Stack Exchange](https://meta.stackexchange.com/questions/384667/font-size-in-the-stacks-editor-when-editing-markdown-is-too-small-please-increa) to request that the size of normal text in the new editor be bumped up to match the old one.

Still, even if the font size is bumped up to match the older editor on the sites, I'd still choose an option to disable syntax highlighting. I find that the text changing as I type it (such as normal text changing to heading format after typing a dash on the next line or text changing to italic format immediately after I type the ending asterisk or underscore) to be extremely jarring. It appears in my eyes as the text flashing, and I'm sensitive to flashing lights (to the point where I had to be guided through a separate staff-only corridor to avoid a mandatory duty-free shop in an airport).

With regard to the second heading, let's take it piece-by-piece:

> Regarding the gray font color used for code (and html), it passes WCAG AAA compliance for small text (7.10:1 in light mode, 8.57:1 in dark mode), which well exceeds the minimum required for accessible contrast in text.

I'm sorry, but I don't find this compelling. On rendered posts on Stack Exchange sites, I've measured the contrast ratio and found it to be exactly 7:1 putting it in compliance with the standard, but I still find them a bit hard to read. Do note that contrast required for readability increases with lower font sizes, and the text in the new editor is smaller than on rendered posts. Perhaps once the linked size issue is implemented, this will no longer be of as much concern, but for other users who still find it too little even after that (perhaps they've changed the contrast setting on their monitor), or if the team declines the linked request, adding an option to disable CommonMark syntax highlighting will resolve it for them.

> In regards to poor italic system fonts, I'm not sure what all we can do there. I'd recommend reaching out to the entity in charge of setting that font on your OS and report your concerns to them.

That's fair. However, in many cases (I'm looking at you, Microsoft), such feedback can take months to be actioned, and in the meantime, there's absolutely nothing a user can do to fix things for themselves (other than user script solutions).

> Regarding text changing as you type it: Is this primarily a concern with the line height / font size of headings, or does it also include color changing as syntax is being recognized?

As I said earlier, it's more to do with the fact that the text visibly flashes when I type certain characters, such as hyphens for headings or the ending characters for bold and italic formatting. I find this extremely jarring, as I note in an above paragraph.

> Regarding blocks that are incorrectly highlighted: Do you have any specific examples of this? If a block is not being properly syntax highlighted, that would be considered a bug.

There is one that I pointed out in the issue report: text formatted as bold and italic using the HTML `<b>` and `<i>` elements and their equivalents `<strong>` and `<em>` aren't highlighted. But just because there may not be any doesn't mean they can't come up in the future, and providing an option to temporarily disable highlighting while the team takes the 6-8 weeks to fix the bug so that the user won't have a broken highlighter for those 6-8 weeks is a good thing.

---

To summarize, I'd strongly prefer an option to disable CommonMark syntax highlighting, both as an accessibility matter and as a matter of personal preference (annoyance).

**Comment by gparyani** [comment]
@b-kelly I know it's been a long holiday period, but it's been two business weeks since I posted the above reply. Just wanted to confirm you've read my above response.

**Comment by b-kelly** [comment]
@gparyani Sorry for the delay. I've been on holiday the past week and spent the week before wrapping things up before going on holiday 😉.

At this point, we are not considering supporting a user facing setting to disable syntax highlighting. This can be somewhat simply fixed with a custom user stylesheet that overrides all the classes listed in the [theme.less](https://github.com/StackExchange/Stacks-Editor/blob/main/src/styles/theme.less) file.

Regarding the text flashing, could you record a small gif of the issue and reply with it here? I'm struggling to understand exactly what the issue is and how that differs from a typical code editor experience. Does this class of issue happen in e.g. VS Code as well?

**Comment by gparyani** [comment]
@b-kelly To be clear, I didn't say that the text *is* flashing; I said it *appears to be flashing* when I type the final character that concludes a formatted block or inline block of text (such as a dash after a heading line or the concluding asterisk for italic formatting). As someone who suffers from a condition that causes my eyes to be hyper-sensitive, my mind interprets it as the text flashing and it feels very jarring. With a typical code editor experience such as Visual Studio Code, the exact same thing happens, but the "typical code editor experience" nearly always has [an option](https://stackoverflow.com/questions/50768252/disable-syntax-highlighting-on-specific-file-in-visual-studio-code) to disable highlighting, which I make use of on files in certain languages (including Markdown).

I strongly suggest you reconsider supporting an option to implement an "accessibility mode" that disables syntax highlighting and also bumps up the size of the text to make it easier for those who are sensitive to apparently flashing text, low font sizes, and low contrast ratios of highlighted text. This option need not be user-facing, but instead triggered by the JS that loads the editor in, and on the Stack Exchange sites, an option can be added to enable editor accessibility mode into the user preferences page.

Also, what about the other parts I mention in my earlier response, and the linked Meta.SE request to increase the editor's font size?

**Comment by gparyani** [comment]
@b-kelly Just letting you know that I'm still waiting for a response to my comment above. I hope it hasn't been forgotten.
